subsys/bluetooth/services/fast_pair/fp_registration_data.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <device.h>
+#include <storage/flash_map.h>
+#include <pm_config.h>
+
+#include "fp_registration_data.h"
+
+#define FP_PARTITION_ID		PM_BT_FAST_PAIR_ID
+#define FP_PARTITION_SIZE	PM_BT_FAST_PAIR_SIZE
+
+static const uint8_t fp_magic[] = {0xFA, 0x57, 0xFA, 0x57};
+
+#define FP_MAGIC_SIZE		(sizeof(fp_magic))
+
+#define FP_DATA_ALIGN		 4
+#define FP_DATA_START		 0
+#define FP_OFFSET(_prev_item_offset, _prev_item_len) \
+	(_prev_item_offset + ROUND_UP(_prev_item_len, FP_DATA_ALIGN))
+
+#define FP_MAGIC_HEADER_OFF	 FP_DATA_START
+#define FP_MODEL_ID_OFF		 FP_OFFSET(FP_MAGIC_HEADER_OFF, FP_MAGIC_SIZE)
+#define FP_ANTI_SPOOFING_KEY_OFF FP_OFFSET(FP_MODEL_ID_OFF, FP_MODEL_ID_LEN)
+#define FP_MAGIC_TRAILER_OFF	 FP_OFFSET(FP_ANTI_SPOOFING_KEY_OFF, FP_ANTI_SPOOFING_PRIV_KEY_LEN)
+
+BUILD_ASSERT(FP_OFFSET(FP_MAGIC_TRAILER_OFF, FP_MAGIC_SIZE) <= FP_PARTITION_SIZE,
+	     "Fast Pair registration data partition is too small");
+
+
+int fp_get_model_id(uint8_t *buf, size_t size)
+{
+	int err;
+	const struct flash_area *fa;
+
+	if (size < FP_MODEL_ID_LEN) {
+		return -EINVAL;
+	}
+
+	err = flash_area_open(FP_PARTITION_ID, &fa);
+	if (!err) {
+		err = flash_area_read(fa, FP_MODEL_ID_OFF, buf, FP_MODEL_ID_LEN);
+		flash_area_close(fa);
+	}
+
+	return err;
+}
+
+int fp_get_anti_spoofing_priv_key(uint8_t *buf, size_t size)
+{
+	int err;
+	const struct flash_area *fa;
+
+	if (size < FP_ANTI_SPOOFING_PRIV_KEY_LEN) {
+		return -EINVAL;
+	}
+
+	err = flash_area_open(FP_PARTITION_ID, &fa);
+	if (!err) {
+		err = flash_area_read(fa, FP_ANTI_SPOOFING_KEY_OFF, buf,
+				      FP_ANTI_SPOOFING_PRIV_KEY_LEN);
+		flash_area_close(fa);
+	}
+
+	return err;
+}
+
+static int validate_fp_magic(const struct flash_area *fa, off_t off)
+{
+	int err;
+	uint8_t read_buf[FP_MAGIC_SIZE];
+
+	err = flash_area_read(fa, off, read_buf, FP_MAGIC_SIZE);
+	if (err) {
+		return err;
+	}
+
+	if (memcmp(read_buf, fp_magic, FP_MAGIC_SIZE)) {
+		err = -EINVAL;
+	}
+
+	return err;
+}
+
+static int fp_registration_data_validate(const struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	int err;
+	const struct flash_area *fa = NULL;
+
+	err = flash_area_open(FP_PARTITION_ID, &fa);
+	if (!err) {
+		err = validate_fp_magic(fa, FP_MAGIC_HEADER_OFF);
+	}
+
+	if (!err) {
+		err = validate_fp_magic(fa, FP_MAGIC_TRAILER_OFF);
+	}
+
+	if (fa) {
+		flash_area_close(fa);
+	}
+
+	__ASSERT(!err, "Invalid content of the Fast Pair partition");
```

Comment:
Added follow-up task for the CRC/hash check

Reply:
I initially assumed that nothing will damage the provisioning data in release configurations. In debug configuration assertion allowed user to quickly realize that something is wrong.

Still, the checks in getters would not hurt. Added.

Reply:
it's more than that. note that these devices may have live span lasting years. It would be great to detect if there is any problem with memory as early as possible and prevent from device usage if memory becomes corrupted.
it can also happens that somebody will try to temper the memory.